src/usb.c

```diff
@@ -52,7 +52,7 @@ ssize_t send_recv(struct stlink_libusb* handle, int terminate,
             3000);
     if (t) {
         printf("[!] send_recv send request failed: %s\n", libusb_error_name(t));
-        return -1;
+        exit(1);
```

Comment:
this solution is not acceptable as stlink can also be used as a library, and a library may never use exit. probably the higher layer should catch the error and exit.

* https://github.com/texane/stlink/blob/master/src/gdbserver/gdb-server.c#L264-L267
* https://github.com/texane/stlink/blob/master/src/gdbserver/gdb-server.c#L1063 
** This function complexity is insane, a lot of `stlink_` calls are not checked for error return values and the main loop should exit not the library itself

Let me know what you think, its a dirty job to fix but its worth it.

Reply:
Does every `-1` response mean `exit` in the higher layer?

Reply:
That is a very good question! IMHO when the libusb backend call fails, indeed there is some sort of failure with the I/O which most of the time you are unable to recover. E.g a stlink instance destroy is nessary, rescan of usb devices and attach again or try with exponential backoff. This is the behaviour `openocd` has when you "hotplug" an stlink programmer which is very nice.

Reply:
Only my opinion: If the backend fails, then it should return an error code (which is "-1").
What I have seen so far, the higher levels will react accordingly to the -1 return value.
So: I would keep this as it is.

Reply:
@xor-gate mentioned an important point here. Therefore I don't see a reason for changing this either.

Reply:
Conversation continues on 21.02.2020.